### PR TITLE
Normalize json dumps

### LINKED ISSUE #1767: Configure json dumps in a single place
Falcon lets the user customize the json serializer used to serialize the media property of a response, but HttpError and SSEvent do not use the user provided serializer, using instead the default python json one.

### PR DESCRIPTION
**Pull request by CaselIT**
# Summary of Changes
Always use the configured media handler for json when loading or serializing it.

# Related Issues
Closes #1767

# Pull Request Checklist

This is just a reminder about the most common mistakes.  Please make sure that you tick all *appropriate* boxes.  But please read our [contribution guide](../CONTRIBUTING.md) at least once; it will save you a few review cycles!

If an item doesn't apply to your pull request, **check it anyway** to make it apparent that there's nothing to do.

- [x] Applied changes to both WSGI and ASGI code paths and interfaces (where applicable).
- [x] Added **tests** for changed code.
- [x] Prefixed code comments with GitHub nick and an appropriate prefix.
- [x] Coding style is consistent with the rest of the framework.
- [x] Updated **documentation** for changed code.
    - [x] Added docstrings for any new classes, functions, or modules.
    - [x] Updated docstrings for any modifications to existing code.
    - [x] Updated both WSGI and ASGI docs (where applicable).
    - [x] Added references to new classes, functions, or modules to the relevant RST file under `docs/`.
    - [x] Updated all relevant supporting documentation files under `docs/`.
    - [x] A copyright notice is included at the top of any new modules (using your own name or the name of your organization).
    - [x] Changed/added classes/methods/functions have appropriate `versionadded`, `versionchanged`, or `deprecated` [directives](http://www.sphinx-doc.org/en/stable/usage/restructuredtext/directives.html?highlight=versionadded#directive-versionadded).
- [x] Changes (and possible deprecations) have [towncrier](https://towncrier.readthedocs.io/en/actual-freaking-docs/index.html) news fragments under `docs/_newsfragments/`, with the file name format `{issue_number}.{fragment_type}.rst`. (Run `towncrier --draft` to ensure it renders correctly.)

If you have *any* questions to *any* of the points above, just **submit and ask**! This checklist is here to *help* you, not to deter you from contributing!

*PR template inspired by the attrs project.*

### CODE REVIEW

File: falcon/app_helpers.py
Comment (vytas7) [inline falcon/app_helpers.py:187]: This is a fair bit of dark magic that IMHO would benefit from a docstring or a `NOTE` explaining how this works in different scenarios: `MEDIA_JSON` is not in `media_handlers`, `MEDIA_JSON` is in media handlers, but it is a custom handler not employing `dumps` etc.

Furthermore, this can potentially lead to hard-to-debug scenarios where the handler does define `dumps`, but the signature is not as we expect here, for instance, it requires an extra indentation level parameter, or what not. We should at minimum discuss this (the special meaning of `dumps`) in the custom handlers docs.

Alternatively, we could be checking the handler type, i.e. if it is strictly of type `JSONHandler`, or a subclass of it (?).

File: docs/_newsfragments/1767.misc.rst
Comment (vytas7) [inline docs/_newsfragments/1767.misc.rst:1]: Towncrier newsfragments are not subject to the docstring conventions; they should IMHO be worded as news items, telling what has happened.

**Review by vytas7 — CHANGES_REQUESTED** [review]
Admittedly, I have some misgiving about the proposed approach.

From a comment above:
> Furthermore, this can potentially lead to hard-to-debug scenarios where the handler does define `dumps`, but the signature is not as we expect here, for instance, it requires an extra indentation level parameter, or what not. We should at minimum discuss this (the special meaning of `dumps`) in the custom handlers docs.

What is more, https://github.com/falconry/falcon/pull/1741 is proposing to add even more ways to handle media, directly (de)serializing from(to) a bytestring as opposed to parsing the request stream.

That distinction (stream/binary string) is probably worth keeping for performance reasons. But maybe we could somehow allow using bytestring handlers for requests too, shimming `stream.read()` where applicable?
The bytestring kind of handler could be also applied not only to WebSockets, but also to SSE, error serialization etc, as also prototyped in this PR.

These are just random thoughts off the top of my head; unfortunately I don't have a one design proposal to rule them all (at least not yet :smiling_imp: )

The way to deprecate stuff at the module level on 3.7+ is cool though, we could actually use it at more places if/when needed.

**Comment by CaselIT** [comment]
> The way to deprecate stuff at the module level on 3.7+ is cool though, we could actually use it at more places if/when needed.

Indeed. It was one of the reason why it was added. [PEP562](https://www.python.org/dev/peps/pep-0562/)

Regarding the rest, I guess we should wait for websocket to land and the update this one.

File: falcon/app_helpers.py
Comment (CaselIT) [reply falcon/app_helpers.py:187]: I think we could just document this in media, specifying that loads and dumps of the json handler if exits and has that attributes are used by the application to handler (de)serialization of json.

**Comment by vytas7** [comment]
One more random thought on unifying serialization. Ideally, the default error serializer should also use response media handlers; so if one is, for instance, building a YAML or Msgpack API, the error structures should by default also pick the default media type set for the API.

Some open questions:
* How could this look for XML?
* How would an error while serializing an exception look like? (See also: https://github.com/falconry/falcon/issues/1607 and [`HTTP 776`](https://github.com/joho/7XX-rfc) )

**Comment by CaselIT** [comment]
This should maybe wait for https://github.com/falconry/falcon/pull/1741

File: falcon/app_helpers.py
Comment (kgriffs) [reply falcon/app_helpers.py:187]: > but the signature is not as we expect here

There are at least three edge cases to address, and to ensure are documented for the developer:

1. A JSON handler is not defined at all. 
2. A JSON handler is defined using ``JSONHandler`` but the dumps/loads require parameters. (We should be OK on this one already since the docs talk about it and recommend ``functools.partial``).
3. A JSON handler is defined but it does not derive from ``JSONHandler``.  In this case it is possible that dumps/loads attributes are defined but not as expected. (The code already falls back to stdlib if dumps/loads are not found... alternatively would it make sense to use the `serialize()` method, although that would require extra branches in the logic since the return type is different?)

TBH I think it would be a more prudent design in the abstract to stick with the ``BaseHandler`` serialize/deserialize interface methods if it doesn't make things too convoluted. That would prevent us from coupling too tightly with the specific ``JSONHandler`` implementation.

File: docs/_newsfragments/1767.misc.rst
Comment (kgriffs) [reply docs/_newsfragments/1767.misc.rst:1]: Capitalize JSON

File: falcon/app_helpers.py
Comment (kgriffs) [inline falcon/app_helpers.py:191]: Just a reminder to make sure the docs point out that this behavior only applies to JSON.

**Review by kgriffs — CHANGES_REQUESTED** [review]

File: falcon/app_helpers.py
Comment (CaselIT) [reply falcon/app_helpers.py]: In the end I've opted to use the handler directly

File: falcon/__init__.py
Comment (CaselIT) [inline falcon/__init__.py:53]: let's support python 4 🚀

File: falcon/asgi/structures.py
Comment (CaselIT) [inline falcon/asgi/structures.py:154]: this is a bit of side effect. encore -> decode -> encode again

File: falcon/http_error.py
Comment (CaselIT) [inline falcon/http_error.py:217]: circular import workaround

File: falcon/request.py
Comment (CaselIT) [inline falcon/request.py:1752]: sadly this is a bit akward

File: tests/test_media_handlers.py
Comment (CaselIT) [inline tests/test_media_handlers.py:212]: I'm not sure if error.to_json should better guard for this case.

**Review by CaselIT — COMMENTED** [review]
I've removed the magic, opting to use the handler directly. But it's has some awkward cases

File: tests/test_media_handlers.py
Comment (CaselIT) [inline tests/test_media_handlers.py:209]: I'm not sure if error.to_json should better guard for this case.

File: falcon/__init__.py
Comment (vytas7) [reply falcon/__init__.py:53]: Lol I already had this type of comment elsewhere :)

File: tests/test_media_handlers.py
Comment (CaselIT) [reply tests/test_media_handlers.py:209]: This was already pointed out by @vytas7 here https://github.com/falconry/falcon/pull/1768#issuecomment-683310866

File: falcon/app.py
Comment (vytas7) [inline falcon/app.py:769]: I cannot find any definition for `MEDIAL_JSON`. Is this a typo of `MEDIA_JSON`?

Comment (vytas7) [inline falcon/app.py:769]: It might also make sense to refer to these constants using the `falcon.` namespace, as otherwise we are implying that the reader has to `from falcon import *`.

File: tests/test_utils.py
Comment (vytas7) [inline tests/test_utils.py:1077]: Great to have extensive testing here. :+1:

File: tests/test_media_handlers.py
Comment (vytas7) [inline tests/test_media_handlers.py:283]: We would automatically wrap sync resources for async routes, as explained on another PR.
So we could DRY this a bit.

File: falcon/asgi/structures.py
Comment (vytas7) [inline falcon/asgi/structures.py:121]: Capitalize JSON here (and elsewhere).

File: falcon/asgi/structures.py
Comment (vytas7) [inline falcon/asgi/structures.py:120]: Capitalize Python, or just remove it altogether IMHO. The reader will know which language we are targeting here.

Comment (vytas7) [inline falcon/asgi/structures.py:120]: Suggested punctuation:
> When not provided, <...>

**Review by vytas7 — CHANGES_REQUESTED** [review]
Some quick documentation nitpicks off the top of my head.
Still trying to wrap my head around all the design implications.

File: tests/asgi/test_sse.py
Comment (vytas7) [inline tests/asgi/test_sse.py:218]: I find it hard to reason about the effects of this `pop` before https://github.com/falconry/falcon/pull/1785 is merged.
But given the test passes, we can probably leave it as-is.

**Review by vytas7 — CHANGES_REQUESTED** [review]

File: falcon/asgi/structures.py
Comment (CaselIT) [reply falcon/asgi/structures.py:121]: I meant the library name, but I'll replace with JSON

File: falcon/http_error.py
Comment (CaselIT) [reply falcon/http_error.py:183]: not sure if it's faster or not. but it's probably easier to read

File: .gitignore
Comment (CaselIT) [reply .gitignore:65]: That's was not supposed to be committed

File: falcon/http_error.py
Comment (CaselIT) [reply falcon/http_error.py:183]: nevermind, I prefer the if version since it's easier to see if both branches are taken. coverage does not consider the or one a branch iirc

File: falcon/asgi/structures.py
Comment (CaselIT) [reply falcon/asgi/structures.py:154]: used return early in this case

File: falcon/http_error.py
Comment (vytas7) [reply falcon/http_error.py:183]: On consideration, I agree.
FWIW, I find the `or` pattern equally easier to read, and more performant. **However,** if a custom media handler is defined as _falsey_ for some strange reason, it would then unexpectedly fallback to the default handler, and it's a surprise we want to avoid.

Comment (CaselIT) [reply falcon/http_error.py:183]: regarding performance there is actually no apparent difference. it's basically a wash.
regarding the falsey it's the same with the if, I'm just doing `not handler`, not `handler is not None` 

personally I don't think it's worth another change :)

Comment (vytas7) [reply falcon/http_error.py:183]: I.e. I prefer `if handler is None` due the falsey object issue.

File: tests/test_media_handlers.py
Comment (CaselIT) [reply tests/test_media_handlers.py:283]: so this would work also without the async resource?

File: falcon/http_error.py
Comment (CaselIT) [reply falcon/http_error.py:183]: ok I guess. A lot of places do use `if obj` or `if not obj` or `obj or other` so with this in mind maybe a more general refactoring is in order?

File: tests/test_media_handlers.py
Comment (CaselIT) [reply tests/test_media_handlers.py]: This probably should be solved in #1607.

File: falcon/http_error.py
Comment (vytas7) [reply falcon/http_error.py:183]: It depends what that `obj` is. If it is something we can control or at least expect a specific type (e.g., primitive built-ins or types defined by us), a boolean test is fine.

**Review by vytas7 — CHANGES_REQUESTED** [review]
Thanks @CaselIT, this is a great effort to unify these scattered handlers, enabling reuse of the same media handler interface.

Unfortunately, it seems that we need to rethink our designs here before we can move forward.

Currently, both your HTTP error serializer and `SSEvent` JSON serializer would simply try to call `handler.serialize(<data>, MEDIA_JSON)`.
However, the ASGI flavour of the framework primarily relies not on `serialize()`, but instead, [`serialize_async()`](https://falcon.readthedocs.io/en/latest/api/media.html#falcon.media.BaseHandler.serialize_async) is `await`ed (although the latter can piggyback on the former, if not defined explicitly).

**Comment by vytas7** [comment]
For reference, reposting some random thoughts of the top of my head from [Gitter](https://gitter.im/falconry/dev?at=5fc29b6cf7e4510fcef788c5):
> I was basically thinking that we would have two types of handlers:
> * Stream handlers, much like the current BaseHandler; the ASGI variant awaits serialization, the WSGI counterpart obviously does not.
> * Binary data handlers, not dissimilar to the current BinaryBaseHandlerWS. Always sync.
>
> The latter would handle WebSocket media, SSE, and error serialization. Error serialization probably being the most questionable one whether it is enough.
> However, in case more complex logic is needed, one could override the awaitable custom error serializer (is it awaitable now?..).

**Comment by CaselIT** [comment]
> 
> 
> Thanks @CaselIT, this is a great effort to unify these scattered handlers, enabling reuse of the same media handler interface.
> 
> Unfortunately, it seems that we need to rethink our designs here before we can move forward.
> 
> Currently, both your HTTP error serializer and `SSEvent` JSON serializer would simply try to call `handler.serialize(<data>, MEDIA_JSON)`.
> However, the ASGI flavour of the framework primarily relies not on `serialize()`, but instead, [`serialize_async()`](https://falcon.readthedocs.io/en/latest/api/media.html#falcon.media.BaseHandler.serialize_async) is `await`ed (although the latter can piggyback on the former, if not defined explicitly).

I've updated to docs to require that the json handler have the sync implementation even for asgi application.

While inconvenient, I don't think it would cause to much of an hassle to users, considering that we also provide a falcon implementation.

**Comment by kgriffs** [comment]
I agree that this approach shouldn't be a big burden on the developer since the differences between the two methods will me small and easy to maintain. Also, most people will just use the default media handlers and not even think about it. 

Noting this requirement in the docs is a good idea. We should also make sure that we raise a user-friendly error if the method is not implemented on the handler.

**Review by kgriffs — CHANGES_REQUESTED** [review]
Just some doc nits. I still need to review the implementation.

**Comment by CaselIT** [comment]
documentation updated @kgriffs

**Comment by vytas7** [comment]
> Noting this requirement in the docs is a good idea. We should also make sure that we raise a user-friendly error if the method is not implemented on the handler.

Could we maybe fallback to the default handler upon `NotImplementedError`, and document this behaviour? It ain't pretty but FAPP it would be quite robust.

**Comment by CaselIT** [comment]
> 
> 
> > Noting this requirement in the docs is a good idea. We should also make sure that we raise a user-friendly error if the method is not implemented on the handler.
> 
> Could we maybe fall back to the default handler upon `NotImplementedError`, and document this behaviour? It ain't pretty but FAPP it would be quite robust.

that may also be an option

File: falcon/media/base.py
Comment (kgriffs) [inline falcon/media/base.py:37]: missing "framework" at the end of this sentence?

**Comment by kgriffs** [comment]
> Could we maybe fall back to the default handler upon NotImplementedError

I kind of think we should just go ahead and raise it to the app just to be sure the developer isn't getting a result they don't expect (i.e., they thought we were calling their thing when all along we weren't but we were silently falling back to something else).

File: falcon/request.py
Comment (CaselIT) [reply falcon/request.py:1752]: @kgriffs should I remove the 'uft-8' here to make it in-line with the other pr?

**Comment by vytas7** [comment]
That's a good point @kgriffs .
I had this as an idea to offer improved damage control, but after all, it's an edge case that shouldn't be seen to often. As such, it's probably better to be explicit indeed.

File: falcon/request.py
Comment (CaselIT) [reply falcon/request.py:1752]: i did

File: tests/test_media_handlers.py
Comment (CaselIT) [reply tests/test_media_handlers.py:270]: i guess that this is what's the env variable `FALCON_ASGI_WRAP_NON_COROUTINES` is for

**Comment by CaselIT** [comment]
updated

File: falcon/request.py
Comment (kgriffs) [inline falcon/request.py:1751]: Yes, this will ding performance, but hopefully it isn't used often enough and the overhead is not great enough to pose a serious problem. But maybe down the road if we unity handling interfaces with the WS handlers by including a deserialize from `bytes` this can be resolved.

**Review by kgriffs — CHANGES_REQUESTED** [review]
LGTM sans that error message nitpick.

File: falcon/asgi/structures.py
Comment (CaselIT) [reply falcon/asgi/structures.py:154]: ah indeed, since they are bytes :)

File: falcon/request.py
Comment (CaselIT) [reply falcon/request.py:1751]: yes, that's an option. I guess this is one thing to consider for v4.

**Comment by CaselIT** [comment]
Updated, I think ready for final review

File: falcon/request.py
Comment (vytas7) [reply falcon/request.py:1751]: I'll try to channel my thoughts into a proposal for 3.x or 4.0: https://github.com/falconry/falcon/pull/1768#issuecomment-738896304

**Review by kgriffs — APPROVED** [review]

File: falcon/media/base.py
Comment (vytas7) [inline falcon/media/base.py:94]: That's actually a neat way to provide a meaningful hint. :+1:

**Review by vytas7 — CHANGES_REQUESTED** [review]
LGTM with one theoretical edge case. Not sure if it is too exotic, or it should be documented.
But if one has such a custom JSON handler that features very sparse output possibly containing `\n\n`, that might break the SSE encapsulation.

Regardless, we also need to restore 100% coverage before merging this. As I understand, a branch is not fully covered here: https://codecov.io/gh/falconry/falcon/pull/1758/diff#D4-2385

**Review by vytas7 — APPROVED** [review]
:+1: (bar the edge case comment above)

**Comment by CaselIT** [comment]
> 
> 
> LGTM with one theoretical edge case. Not sure if it is too exotic, or it should be documented.
> But if one has such a custom JSON handler that features very sparse output possibly containing `\n\n`, that might break the SSE encapsulation.

oh, that's quite an edge case. `print(json.dumps({'x':'a'}, indent='\n\n\n'))`
Should we strip `\n\n` from the returned object?

**Comment by vytas7** [comment]
Maybe a documentation note is enough?

**Pull request merged**
